drivers: modem: ublox-sara-r4: add TLS offload support

johann2 opened the pull request:
Currently it's able to connect to google iot mqtt server. Tested in our private application.
All other use cases are untested.
The only related test to this driver, `drivers/build_all` compiles at least.

I'm not exactly proud of changing the `modem_socket` protocol type to int, but it was the simplest way that needed the least changes elsewhere.

Signed-off-by: Johann Tael <jntael@gmail.com>

johann2 committed:
drivers: modem: ublox-sara-r4: add TLS offload support

Currently it's able to connect to google iot. All other use cases are
untested.

Signed-off-by: Johann Tael <jntael@gmail.com>

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -437,6 +443,81 @@ static ssize_t send_socket_data(void *obj,
+			 int cert_type)
+{
+	int ret;
+	char filenames[3][10] = {
```
There is some RAM wastage with this approach, how about:
```
	char *filenames[] = {
		"ca",
		"cert",
		"pkey"
	};
```

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1258,6 +1348,41 @@ static int create_socket(struct modem_socket *sock, const struct sockaddr *addr)
+		snprintk(buf, sizeof(buf), "AT+USECPRF=%d", sock->id);
+		modem_cmd_send(&mctx.iface, &mctx.cmd_handler, NULL, 0U, buf, &mdata.sem_response,
+			       MDM_CMD_TIMEOUT);
+		/* Disable server cert validation TODO:Validation level should be at least 1*/
```
This looks like a serious issue to me, even for initial implementation. According to the documentation I've found, you don't need to register any root CA at all to successfully complete the handshake as the server certificate is ignored in this case. If we want to get this through, IMO it should be listed somewhere as a current limitation **in bold**.

johann2 replied:
I've updated the code to set validation level to 1, which means according to manual:
> Level 1: certificate validation against a specific or a list of imported trusted root certificates.
Ideally, level 2 / 3 should be supported as well, but I can't afford to put more work into this right now.

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1258,6 +1348,41 @@ static int create_socket(struct modem_socket *sock, const struct sockaddr *addr)
+		modem_cmd_send(&mctx.iface, &mctx.cmd_handler, NULL, 0U, buf, &mdata.sem_response,
+			       MDM_CMD_TIMEOUT);
+		/* Set root CA filename */
+		snprintk(buf, sizeof(buf), "AT+USECPRF=%d,3,\"ca\"", sock->id);
```
This looks like another limitation - only a single root CA is supported, while the TLS sockets allow configuring PSK security as well. Should be listed somewhere as well.

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1639,6 +1764,92 @@ static ssize_t offload_sendmsg(void *obj, const struct msghdr *msg, int flags)
+		tag = sec_tags[i];
+		cert = credential_next_get(tag, NULL);
+		while (cert != NULL) {
+			/* Map Zephyr cert types to Simplelink cert options: */
```
Simplelink?

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1639,6 +1764,92 @@ static ssize_t offload_sendmsg(void *obj, const struct msghdr *msg, int flags)
+			case TLS_CREDENTIAL_CA_CERTIFICATE:
+				id = 0;
+				break;
+			case TLS_CREDENTIAL_SERVER_CERTIFICATE:
```
Will server certificate/private key pair even work if configured? It seems that the implementation is fixed to use CA cert right now.

johann2 replied:
This was dead code left in during develompent. I've removed the handling for all certs except for the CA to avoid any confusion

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1639,6 +1764,92 @@ static ssize_t offload_sendmsg(void *obj, const struct msghdr *msg, int flags)
+			break;
+		case TLS_HOSTNAME:
+		case TLS_PEER_VERIFY:
+			ret = 0;
```
This doesn't rook right, if it's not implemented then report it back (-ENOTSUP).

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -437,6 +443,82 @@ static ssize_t send_socket_data(void *obj,
+					    NULL, 0U, false);
+	k_sem_give(&mdata.cmd_handler_data.sem_tx_lock);
+
+	if (ret < 0) {
```
Just simply `return ret`, it should be 0 if there were no errors.

rlubos replied:
Still not addressed.

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1258,6 +1349,41 @@ static int create_socket(struct modem_socket *sock, const struct sockaddr *addr)
+		char buf[sizeof("AT+USECPRF=#,#,#######\r")];
+
+		snprintk(buf, sizeof(buf), "AT+USOSEC=%d,1,%d", sock->id, sock->id);
+		modem_cmd_send(&mctx.iface, &mctx.cmd_handler, NULL, 0U, buf, &mdata.sem_response,
```
Would be good to verify the return value of `modem_cmd_send()`, I assume that failed transition of any of these commands results in an invalid socket to use.

rlubos reviewed (commented):
Still a few shortcomings. I'm no maintainer of this driver, but I guess this code be used as a minimal starting point for TLS implementation to the driver.

johann2 commented:
Addressed all the comments and removed unneeded code that disables security on non-TLS sockets as the security is disabled by default on socket creation.

rlubos commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -437,6 +443,77 @@ static ssize_t send_socket_data(void *obj,
+	k_sleep(MDM_PROMPT_CMD_DELAY);
+	mctx.iface.write(&mctx.iface, cert_data, cert_len);
+
+	k_timeout_t timeout = Z_TIMEOUT_MS(1000);
```
This variable is no longer needed.
